### ISSUE TITLE
kvs: potential minor performance improvements                                                                                                     

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#3615]
While doing profiling analysis for #3583, a few minor KVS hotspots showed up that could be  improved:

```
                  - 2.49% disconnect_request_cb
                     - 2.49% cache_wait_destroy_msg
                        - 1.49% wait_destroy_msg
                             1.14% zlist_first
                          0.70% fzhashx_next
```

Dunno if some trickery could be used to eliminate the need to call `FOREACH_ZHASHX`.  Hypothetically all entries with messages could be hashed to limit the iteration.  Or perhaps a list of entries with messages could be on a list, to limit the length of iteration.

```
         - 2.03% transaction_check_cb
            - 2.03% kvsroot_mgr_iter_roots
               - 1.94% kvstxn_check_root_cb
                  - 1.84% kvstxn_apply
                     - 1.62% kvstxn_process
                          0.72% kvstxn_link_dirent
                        - 0.68% kvstxn_unroll
                           - 0.61% kvstxn_unroll
                                0.55% kvstxn_unroll
```

the `kvsroot_mgr_iter_roots` could maybe be improved with some list or something that indicates which roots have transactions ready instead of just blindly iterating through all roots all the time.  With high throughput workflows, there's going to be thousands of namespaces, being iterated through all of the time.

**Comment by chu11** [comment]
doing another profile run (using simulated execution):

```
         - 1.18% transaction_prep_cb 
            - kvsroot_mgr_iter_roots
                 0.72% kvstxn_prep_root_cb    
```

could try to have some flag or counter indicating a "transaction is available" to avoid another unnecessary looping.

**Comment by chu11** [comment]
Started to look into this.  I actually had little idea where to start, so ran `flux mini run hostname` on TOSS3 and TOSS4 and looked at the exec eventlog.

TOSS3

```
1619364043.471757 init
1619364043.473668 starting
1619364043.524205 shell.init leader-rank=0 size=1 service="8556-shell-64441286656"
1619364043.527429 shell.start task-count=1
1619364043.532992 complete status=0
1619364043.533042 done
```

TOSS4

```
1619364043.754561 init
1619364043.756117 starting
1619364043.859735 shell.init leader-rank=0 size=1 service="8556-shell-46489665536"
1619364043.862561 shell.start task-count=1
1619364043.867620 complete status=0
1619364043.867658 done
```

hmm, this was unexpected.  Well it ends up the slowdown appears to only exist when flux is under a tad of load.

Example when I run 1 job by itself, then 16 via the throughput test.

```
>flux mini run hostname
>./throughput_test.sh 16 1
```

TOSS3

```
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME NODELIST
    f84bJagY achu     true       CD      1      1   0.378s catalyst160
    f84bJagX achu     true       CD      1      1   0.383s catalyst160
    f84WrcqW achu     true       CD      1      1   0.402s catalyst160
    f84YLc7r achu     true       CD      1      1   0.396s catalyst160
    f84ZpbQD achu     true       CD      1      1   0.382s catalyst160
    f84bJagZ achu     true       CD      1      1   0.355s catalyst160
    f84cnZxs achu     true       CD      1      1   0.355s catalyst160
    f84WrcqX achu     true       CD      1      1   0.379s catalyst160
    f84ZpbQC achu     true       CD      1      1   0.371s catalyst160
    f84WrcqV achu     true       CD      1      1   0.382s catalyst160
    f84VNdZA achu     true       CD      1      1   0.381s catalyst160
    f84VNdZ9 achu     true       CD      1      1   0.354s catalyst160
    f84YLc7q achu     true       CD      1      1   0.336s catalyst160
    f84TteGp achu     true       CD      1      1   0.292s catalyst160
    f84TteGo achu     true       CD      1      1   0.281s catalyst160
    f2geKhWF achu     hostname   CD      1      1   0.062s catalyst160
```

TOSS4

```
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME NODELIST
    ƒ6aezDFE achu     true       CD      1      1   1.387s opal186
    ƒ6ahxBox achu     true       CD      1      1   1.362s opal186
    ƒ6ajSB6F achu     true       CD      1      1   1.234s opal186
    ƒ6ajSB6G achu     true       CD      1      1   1.233s opal186
    ƒ6ahxBou achu     true       CD      1      1   1.229s opal186
    ƒ6agUCXd achu     true       CD      1      1   1.157s opal186
    ƒ6agUCXc achu     true       CD      1      1   1.150s opal186
    ƒ6ahxBow achu     true       CD      1      1   1.058s opal186
    ƒ6agUCXa achu     true       CD      1      1   1.074s opal186
    ƒ6aezDFD achu     true       CD      1      1   1.072s opal186
    ƒ6ahxBov achu     true       CD      1      1   0.927s opal186
    ƒ6ahxBoy achu     true       CD      1      1   0.927s opal186
    ƒ6agUCXb achu     true       CD      1      1   0.941s opal186
    ƒ6aezDFF achu     true       CD      1      1   0.940s opal186
    ƒ6agUCXZ achu     true       CD      1      1   0.795s opal186
    ƒ6aezDFG achu     true       CD      1      1   0.779s opal186
    ƒ47pXQWK achu     hostname   CD      1      1   0.098s opal186
```

notice how the runtime is a lot larger under TOSS4, max 1.387 seconds vs max .402 seconds on TOSS3.

and sure enough, the shell startup now takes time up about 1.3 seconds.

```
>flux job eventlog -p guest.exec.eventlog ƒ6aezDFE
1619389519.314715 init
1619389519.338146 starting
1619389520.691006 shell.init leader-rank=0 size=1 service="8556-shell-212433108993"
1619389520.693714 shell.start task-count=1
1619389520.700053 complete status=0
1619389520.700114 done
```

I had little idea how to proceed, so I decided to profile and see if I can find something interesting.  Basically ran:

`perf record -g --call-graph=dwarf -D 60000 src/cmd/flux start ./throughput_loop.sh 1024 1`

basically profile running 1024 jobs being run

The big thing I noticed under TOSS4 compared to TOSS3 (only listing `lt-flux-shell` stuff):

```
+    7.85%     0.00%  lt-flux-shell    ld-2.28.so                                     [.] _dl_init
+    7.85%     0.00%  lt-flux-shell    ld-2.28.so                                     [.] call_init.part.0
+    7.85%     0.00%  lt-flux-shell    ld-2.28.so                                     [.] _dl_start_user
+    7.78%     0.00%  lt-flux-shell    libcrypto.so.1.1.1g                            [.] 0x00001555512f206f
+    7.65%     0.00%  lt-flux-shell    libcrypto.so.1.1.1g                            [.] 0x00001555513eecfe
+    7.65%     0.00%  lt-flux-shell    libcrypto.so.1.1.1g                            [.] FIPS_selftest
+    6.75%     0.00%  lt-flux-shell    lt-flux-shell                                  [.] main
+    6.75%     0.00%  lt-flux-shell    libc-2.28.so                                   [.] __libc_start_main
+    6.74%     0.00%  lt-flux-shell    lt-flux-shell                                  [.] _start
+    5.16%     0.70%  lt-flux-shell    ld-2.28.so                                     [.] _dl_relocate_object
+    4.80%     0.87%  lt-flux-shell    ld-2.28.so                                     [.] _dl_lookup_symbol_x
+    4.50%     0.00%  lt-flux-shell    lt-flux-shell                                  [.] shell_init (inlined)
+    4.40%     0.01%  lt-flux-shell    libcrypto.so.1.1.1g                            [.] BN_mod_exp_mont_consttime
```

lib `ld` and `libcrypto` are using up way more cycles under TOSS4 than TOSS3.

A selection of profiling call trees


```
-    7.85%     0.00%  lt-flux-shell    ld-2.28.so                                     [.] _dl_start_user
     _dl_start_user
   - _dl_init
      - call_init.part.0
         - 7.78% 0x1555512f206f
            - 7.65% 0x1555513eecfe
               - FIPS_selftest
                  - 1.88% 0x1555513f4d84
                     - 1.84% 0x1555513f45ee
                        - EVP_DigestSignFinal
                           - 1.84% 0x155551438561
                              - RSA_sign
                                 - 1.10% 0x15555143647c
                                    - 0.52% 0x155551434ead
                                         BN_mod_exp_mont_consttime
                                    - 0.52% 0x155551434e49
                                         0.52% BN_mod_exp_mont_consttime
                                 - 0.71% 0x15555143655f
                                    - 0.71% 0x1555514351d2
                                         RSA_setup_blinding
                                         BN_BLINDING_create_param
                  - 1.19% 0x1555513f3212
                     - 0.63% 0x1555513f45ee
                        - EVP_DigestSignFinal
                           - 0.63% 0x15555136ed06
                              - DSA_sign
                                 - 0.59% 0x15555136e55d
                                    - 0.53% 0x15555136e3ab
                                         BN_mod_exp_mont_consttime
                     - 0.56% 0x1555513f47e4
                        - EVP_DigestVerifyFinal
                           - 0.56% DSA_verify
                              - 0.51% 0x15555136dde4
                                   BN_mod_exp2_mont
                  - 1.13% 0x1555513f4eb8
                     - 1.11% 0x155551435d1c
                        - 0.55% 0x155551434ead
                             0.55% BN_mod_exp_mont_consttime
                        - 0.54% 0x155551434e49
                             0.53% BN_mod_exp_mont_consttime
                  - 1.12% 0x1555513f4dc1
                     - 1.08% 0x1555513f45ee
                        - EVP_DigestSignFinal
                           - 1.08% 0x155551438436
                              - 1.07% 0x15555143647c
                                 - 0.52% 0x155551434e49
                                      BN_mod_exp_mont_consttime
                                 - 0.51% 0x155551434ead
                                      BN_mod_exp_mont_consttime
                  - 0.82% 0x1555514275cf
                     - ERR_set_mark
                        - ERR_get_state
                           - 0.81% OPENSSL_init_crypto
                              - CRYPTO_THREAD_run_once
                              - __pthread_once_slow
                                   0.81% 0x1555513f7c0c
                  - 0.54% 0x1555513ef95d
                     - 0.53% 0x155551367b55
                          BN_mod_exp_mont_consttime
```

```
-    6.75%     0.00%  lt-flux-shell    lt-flux-shell                                  [.] main
   - 6.75% main
      - 4.50% shell_init (inlined)
         - 3.90% plugstack_call
            - 3.89% flux_plugin_call
               - 3.56% affinity_init
                  - 3.55% shell_affinity_create (inlined)
                     - 3.54% shell_affinity_topology_init (inlined)
                        - 2.11% hwloc_topology_load
                           - 1.98% 0x155553a5268d
                                1.10% 0x155553a8729e
                        - 1.42% 0x155553a56afc
                           - 1.37% 0x155553a62b86
                              - 1.37% __dlopen_check (inlined)
                                 - _dlerror_run
                                    - 1.37% __GI__dl_catch_error (inlined)
                                       - __GI__dl_catch_exception (inlined)
                                          - 1.37% dlopen_doit
                                             - _dl_open
                                                - 1.35% __GI__dl_catch_exception (inlined)
                                                   - 1.35% dl_open_worker
                                                      - 1.15% _dl_relocate_object
                                                         - 1.02% _dl_lookup_symbol_x
                                                              do_lookup_x

```

dynamic loading seems to be eating up a lot more cycles compared to TOSS3.  There's a lot of openssl calls going on in there.  I don't know exactly what the functions in the call tree do, but I imagine anything called "Digest" or "Sign" is not relatively fast and doing it through 1024 jobs is going to eat up time.

A theory I came up out of thin air, could Redhat have enabled some special signing of libs / modules / etc. in RHEL8 compared to RHEL7?  For the average program on a non-busy system it's no big deal, but with hundreds of jobs being run simultaneously (such as their `make -j16 check` or our throughput tests), things just get bogged down.  

I wasn't able to find anything out online and couldn't find any smoking gun on RHEL8 release notes.

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/8.0_release_notes/index#security

But for all I know it could be configuration that was possible on RHEL7 and we just didn't enable before.

Perhaps something to ping SAG team or Redhat folks about when I come up for air again.

**Comment by chu11** [comment]
closing, as #3633 will overtake this

**Comment by chu11** [comment]
re-opening, the transaction prep & check will be handled via #3633, but `disconnect_request_cb` / `cache_wait_destroy_msg` isn't really related to that one.

**Comment by chu11** [comment]
doing some light code skimming the `disconnect_request_cb` & `cache_wait_destroy_msg`  path could be optimized half easily using the ccan list implementation and some type of `waitlist_valid` and `waitlist_notdirty` list.  Not urgent since we don't have a benchmark heavily hammering this path, but if we need to eek out some extra cycles, an easy one to add.

**Comment by chu11** [comment]
another medium bottleneck in the KVS to consider (using @grondo nsbench, percentages should be ignored compared to the above).

```
         - 18.26% content_store_completion                                                                                                 ▒
              cache_entry_set_dirty                                                                                                        ▒
              wait_runqueue                                                                                                                ▒
            - wait_runone (inlined)                                                                                                        ▒
               - 18.26% kvstxn_apply                                                                                                       ▒
                  - 18.10% kvstxn_process                                                                                                  ▒
                     - keys_from_ops (inlined)                                                                                             ▒
                        - 17.97% normalize_and_append_unique (inlined)                                                                     ▒
                             8.86% __strcmp_sse42                                                                                          ▒
                             2.58% json_string_value                                                                                       ▒
                             1.03% strcmp@plt                                                                                              ▒
                             0.97% json_string_value@plt                                                                                   ▒
                             0.96% json_array_get@plt                                                                                      ▒
                             0.95% json_array_size@plt          
```

the `normalize_and_append_unique()` is a little heavy.  It calls `kvs_util_normalize_key()` which does this:

```
    char *cpy = strdup (key);
    int i, len = strlen (key) + 1;
    bool has_sep_suffix = false;

    if (cpy) {
        /* Transform duplicate path separators into a single one.
         */
        for (i = 0; i < len; ) {
            if (cpy[i] == sep && cpy[i + 1] == sep) {
                memmove (&cpy[i], &cpy[i + 1], len - i - 1);
                len--;
            }
            else
                i++;
        }
```

So there's an `strdup()` for every key and there's an iteration through every char of every key.  There might be an easy solution here.  We could punt normalization onto `libkvs` and the client side, so that it's not all bottle-necked on the broker.

Then in `normalize_and_append_unique()` we have this:

```
    json_array_foreach (keys, index, value) {
        const char *s = json_string_value (value);
        if (s && !strcmp (s, key_norm)) {
            unique = false;
            break;
        }
    }
```

so O(N^2) iterating through every key in the each setroot event, doing an `strcmp()`, just to make sure no duplicates.

perhaps could use a `json_object()` instead of array?  Internally there's a hash that can do the lookup faster?

**Comment by chu11** [comment]
This one is really minor, but since we're at it, we might want to consider also:

```
               - 0.63% sync_cb                                                                                                             ▒
                    0.61% cache_expire_entries  
```

cache expire entries does a pretty expensive `zhashx_keys()` call, which goes through every bucket & entry in the hash.  Then makes a `zlistx` out of those keys to return to the user.

A "lru" or similar of current cache entries might be reasonable.  But of course we're looking at something that eats up 0.63% of time.  Could investigate if this is worth it.

**Comment by chu11** [comment]
played around with a ccan list to improve `sync_cb` iteration.  Since `sync_cb()` is only called periodically based on heartbeats and we're including additional operations in the common path (`cache_insert()`, `cache_remove()`), performance improvement was not discernibly better.  Perhaps someday it could be if we up the heartbeat.  For the time being, I'll just park this branch here:

https://github.com/chu11/flux-core/tree/issue3615_kvs_sync_tryC

**Comment by garlick** [comment]
Might want to post it as a PR anyway - if it cleans things up a bit it might be worth pulling in even if it doesn't change much for this test. There might be other benefits to reducing malloc usage and not blocking the KVS thread with scans that we're not drawing out in test.

I'm guessing it's pretty safe to change since you've put so much into the unit tests for the KVS internals.

**Comment by chu11** [comment]
Gonna go ahead and close this as #3654 #3658 #3663 are merged and #3664 is an optional PR to merge.

**Comment by chu11** [comment]
Just posting in case i ever want to look back at this.  I experimented with this fix as well.

```
diff --git a/src/modules/kvs/kvstxn.c b/src/modules/kvs/kvstxn.c
index 5114e2c..70aa845 100644
--- a/src/modules/kvs/kvstxn.c
+++ b/src/modules/kvs/kvstxn.c
@@ -294,7 +294,7 @@ static int store_cache (kvstxn_t *kt, json_t *o,
         }
     }
     else {
-        if (treeobj_validate (o) < 0 || !(data = treeobj_encode (o))) {
+        if (!(data = treeobj_encode (o))) {
             flux_log_error (kt->ktm->h, "%s: treeobj_encode", __FUNCTION__);
             goto error;
         }
@@ -384,6 +384,10 @@ static int kvstxn_unroll (kvstxn_t *kt, json_t *dir)
             }
             if (!(ktmp = treeobj_create_dirref (ref)))
                 return -1;
+            if (treeobj_validate (ktmp) < 0) {
+                json_decref (ktmp);
+                return -1;
+            }
             if (json_object_iter_set_new (dir, iter, ktmp) < 0) {
                 json_decref (ktmp);
                 errno = ENOMEM;
@@ -405,6 +409,10 @@ static int kvstxn_unroll (kvstxn_t *kt, json_t *dir)
                 }
                 if (!(ktmp = treeobj_create_valref (ref)))
                     return -1;
+                if (treeobj_validate (ktmp) < 0) {
+                    json_decref (ktmp);
+                    return -1;
+                }
                 if (json_object_iter_set_new (dir, iter, ktmp) < 0) {
                     json_decref (ktmp);
                     errno = ENOMEM;
```


`treeobj_validate()` will recursively validate all entries in a directory.  So if a directory were extremely large, it could eat up cycles.  The idea behind this was to move `treeobj_validate()` to validate only "new" things added to a directory.  Assuming that all prior entries were validated beforehand.

Experiments show no benefit at the moment.  This one is highly dependent on the data that exists in the KVS, vs. other fixes that were general improvements (i.e. lower iteration on stuff generally benefits all benchmarks).

So not going to pursue this in a PR, b/c it can be a tad confusing what's going on.  But just wanted to note this for the future for reconsideration.
